### PR TITLE
Add matmul ONNX op support

### PR DESCRIPTION
**Pull request by laggui**
### Related Issues/PRs

Progress towards #1544 

### Changes

Added matmul ONNX op import support and marked mul (already supported).

### Testing

Added unit tests for codegen and expected outputs

### CODE REVIEW

File: crates/burn-import/onnx-tests/tests/matmul/matmul.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    # Export to onnx
+    model = Model()
+    model.eval()
+    device = torch.device("cpu")
+    onnx_name = "matmul.onnx"
+    a = torch.arange(24, dtype=torch.float, device=device).reshape(1,2,3,4)
+    b = torch.arange(16, dtype=torch.float, device=device).reshape(1, 2, 4, 2)
+    # b = torch.ones(1, 2, 4, 2, device=device)
+    # a = torch.randn(1, 2, 3, 4, device=device)
+    # b = torch.randn(1, 2, 4, 2, device=device)
```
Comment (nathanielsimard) [inline crates/burn-import/onnx-tests/tests/matmul/matmul.py:30]: Dead code?

File: crates/burn-import/src/onnx/dim_inference.rs [hunk 2]
```diff
@@ -423,3 +424,12 @@ fn conv_transpose2d_update_outputs(node: &mut Node) {
...
     }
 }
+
+fn matmul_update_outputs(node: &mut Node) {
+    // NOTE: matmul only supported for float tensors (no broadcasting support)
+    if let ArgType::Tensor(tensor) = node.inputs[0].clone().ty {
+        node.outputs[0].ty = ArgType::Tensor(tensor);
+    } else {
+        panic!("Only tensor input is valid");
+    }
```
Comment (nathanielsimard) [inline crates/burn-import/src/onnx/dim_inference.rs:434]: We are supposed to support broadcasting, but we have to be explicit and call unsqueeze during codegen.

**Review by antimora — CHANGES_REQUESTED** [review]
Thanks for adding this OP. 

I think we have opportunity here to fix the broadcasting issue as well. 

Burn supports a partial broadcasting (see my issue regarding this: https://github.com/tracel-ai/burn/issues/1499) where if one of the dimensions is 1, then it's broadcasted. But broadcasting won't work if dimensions do not match (2 vs 1) because Burn cannot dynamically expand static constant type (D). However, there is an easy workaround the user can apply (as I mentioned in https://github.com/tracel-ai/burn/issues/1499) - namely use `unsqueeze` operations. But the user needs to know which dimension is greater. So we can do the same here since we are just generating rust code. We just need to know if dimensions do not match, we add `unsqueeze`.

I hope it's easy change for you but with this we will be 100% ONNX compatible. The same logic can be used for other ops such as addition.

File: crates/burn-import/src/onnx/dim_inference.rs (hunk 2 above)
Comment (antimora) [reply crates/burn-import/src/onnx/dim_inference.rs:434]: Haha I was writing the same. See my comment below.

File: crates/burn-import/onnx-tests/tests/matmul/matmul.py (hunk 1 above)
Comment (laggui) [reply crates/burn-import/onnx-tests/tests/matmul/matmul.py:30]: Dead code strikes again 😄

**Comment by laggui** [comment]
Oh it wasn't clear even to me that this type of broadcasting was supported lol.

Will fix! Thanks for the comments.

**Comment by antimora** [comment]
> Oh it wasn't clear even to me that this type of broadcasting was supported lol.
> 
> Will fix! Thanks for the comments.

Same! That's why I have added this ticket (https://github.com/tracel-ai/burn/issues/1499) to document it in details. I think many will be confused by it too.

**Review by nathanielsimard — APPROVED** [review]
LGTM

**Review by antimora — APPROVED** [review]
Thank you for completing this task and making matmul broadcastable in broadest sense. 

Also thanks for explaining difference case scenarios offline to me. We definitely need a documentation to explain Burn's broadcasting rules.

**Pull request merged**